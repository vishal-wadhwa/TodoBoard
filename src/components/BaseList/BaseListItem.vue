<template>
  <v-container fluid class='pa-1' @click='onItemClickHandler'>
    <v-card v-if='!isNewView'>
      <v-layout
        row
        align-start
        :style='{borderLeftColor: ribbonColor}'
        class='pa-2 bli-ribbon-style'
      >
        <slot name='vf-icon' :type='type'>
          <v-icon>{{type}}</v-icon>
        </slot>
        <v-flex align-self-center class='mt-1 mx-2'>
          <div class='title font-weight-regular'>
            <slot name='vf-title' :title='title' :view='view'>
              <v-card-title class='pa-0'>{{title}}</v-card-title>
            </slot>
          </div>
          <v-divider class='my-2' v-if='!isCompactView'></v-divider>
          <v-card-text text-truncate class='ml-1 pa-0' v-if='!isCompactView'>{{desc}}</v-card-text>
          <v-card-text class='pa-0' v-if='!isCompactView'>
            <slot name='tags' :title='title' :tags='tags'>
              <v-chip
                v-for='(tag, idx) in tags'
                :key='idx'
                small
                label
                :color='tagColor()'
                text-color='white'
              >{{tag}}</v-chip>
            </slot>
          </v-card-text>
        </v-flex>
      </v-layout>
    </v-card>
    <slot v-else name='vn-body' :title='title' :view='view'>
      <v-card flat tile>
        <v-card-actions class='bli-new--border pa-0'>
          <v-btn flat depressed block class='bli-new--height px-0'>
            <slot name='vn-content' :title='title' :view='view'>
              <v-icon>note_add</v-icon>
              {{title}}
            </slot>
          </v-btn>
        </v-card-actions>
      </v-card>
    </slot>
  </v-container>
</template>
<script>
const FULL_VIEW = 'full'
const COMPACT_VIEW = 'compact'
const NEW_VIEW = 'new'
const VIEW_TYPES = [FULL_VIEW, COMPACT_VIEW, NEW_VIEW]
export default {
  data () {
    return {
      colorChoices: ['red', 'green', 'purple', 'indigo', 'primary', 'secondary', 'teal'],
      FULL_VIEW: FULL_VIEW,
      COMPACT_VIEW: COMPACT_VIEW,
      VIEW_TYPES: VIEW_TYPES,
      NEW_VIEW: NEW_VIEW
    }
  },
  props: {
    type: {
      type: String,
      default: 'assignment'
    },
    title: {
      type: String,
      required: true
    },
    desc: {
      type: String
    },
    view: {
      type: String,
      default: NEW_VIEW,
      validator (val) {
        return VIEW_TYPES.includes(val)
      }
    },
    ribbonColor: {
      type: String,
      default: 'black'
    },
    tags: {
      type: Array,
      default: () => []
    }
  },
  methods: {
    tagColor () {
      const idx = Math.random() * this.colorChoices.length >> 0
      return this.colorChoices[idx]
    },
    onItemClickHandler (ev) {
      this.$emit('bli:click', { ...this.$props }, ev)
    }
  },
  computed: {
    isCompactView () {
      return this.view === this.COMPACT_VIEW || ((!this.tags || this.tags.length === 0) && !this.desc)
    },
    isNewView () {
      return this.view === this.NEW_VIEW
    }
  }
}
</script>
<style lang="stylus" scoped>
.bli-new {
  &--border {
    border: 2px dotted gray;
  }

  &--height {
    height: 40px;

    @media screen and (min-width: 960px) {
      height: 80px;
    }
  }
}

.bli-ribbon-style {
  border-left: 3px solid black;
}
</style>
