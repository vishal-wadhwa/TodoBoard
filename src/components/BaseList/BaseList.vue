<template>
  <v-container fluid>
    <div class='headline mb-3 font-weight-light'>
      {{header}}
      <v-avatar :color='highlightColor' size='30' class='white--text ml-3'>{{dataList.length}}</v-avatar>
    </div>
    <base-list-item
      @bli:click='(data, ev) => onItemClickHandler(item._id, data, ev)'
      v-for='item in dataList'
      :key='item._id'
      :title='item.title'
      :desc='item.desc || ""'
      :tags='item.tags || []'
      :ribbon-color='highlightColor'
      :view='item.view || "full"'
    ></base-list-item>
    <base-list-item view='new' title='Create new' @bli:click='onItemAddHandler'></base-list-item>
  </v-container>
</template>
<script>
import BaseListItem from './BaseListItem'

export default {
  components: { BaseListItem },
  data () {
    return {
      dataList: []
    }
  },
  props: {
    highlightColor: {
      type: String,
      default: 'black'
    },
    showCount: {
      type: Boolean,
      default: true
    },
    header: {
      type: String,
      required: true
    },
    list: {
      type: Array,
      default: () => []
    }
  },
  methods: {
    onItemAddHandler (data, ev) {
      this.$emit('bl:add', ev)
    },
    onItemClickHandler (id, data, ev) {
      this.$emit('bl:click', this.list.find(it => it['_id'] === id), ev)
    }
  },
  computed: {

  },
  created () {
    this.dataList = this.list.slice()
  },
  watch: {
    list (newv, oldv) {
      this.dataList = newv.slice()
    }
  }
}
</script>
<style lang="stylus" scoped>
</style>
