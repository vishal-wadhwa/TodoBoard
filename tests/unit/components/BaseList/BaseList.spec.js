import { shallowMount, createLocalVue, mount } from '@vue/test-utils'
import BaseList from '@/components/BaseList/BaseList'
import BaseListItem from '@/components/BaseList/BaseListItem'
import Vuetify from 'vuetify'
import Vue from 'vue'
Vue.use(Vuetify)

describe('BaseList.vue', () => {
  let localVue = null
  const factory = (opts = {}, full = false) => {
    return full ? mount(BaseList, opts) : shallowMount(BaseList, opts)
  }
  beforeEach(() => {
    localVue = createLocalVue()
    // localVue.use(Vuetify)
  })

  const header = 'List1'
  const showCount = false
  const list = [{
    'title': 'sanmu',
    'desc': 'oh cunihnon vocew lab sirrufoka loj ujo utceb lez ej',
    'tags': ['fa', 'jebpu'],
    '_id': 87,
    'type': 'list'
  }, {
    'title': 'iz',
    'desc': 'vuha ogpufub in jufwozi mira ep po gozseic burni zictam',
    'tags': ['sikgaok', 'co'],
    '_id': 21,
    'type': 'playlist_play'
  }, {
    'title': 'duzjukve',
    'desc': 'giv howseb ikopir lo vehladdih delolir uj dafassu joig mocun',
    'tags': ['magew', 'wobsudig'],
    '_id': 93
  }]

  it('renders list header', () => {
    const wrapper = factory({
      propsData: { header },
      localVue
    })
    expect(wrapper.text()).toMatch(header)
  })

  it('shows count by default', () => {
    const wrapper = factory({
      propsData: { header },
      localVue
    })
    const avatarWrapper = wrapper.find('div.headline :first-child')
    expect(avatarWrapper.text()).toMatch('0')
  })

  it('does not show count when prop set to `false`', () => {
    const wrapper = factory({
      propsData: { showCount, header },
      localVue
    })
    const avatarWrapper = wrapper.find('div.headline :first-child')
    expect(avatarWrapper.exists()).toBeFalsy()
  })

  it('renders only 1 new list item when list is not passed', () => {
    const wrapper = factory({
      propsData: { header }
    })
    const baseListItems = wrapper.findAll(BaseListItem)
    expect(baseListItems.length).toEqual(1)
    expect(baseListItems.at(0).vm.$props['view']).toEqual('new')
  })

  it('renders a list of correct length and count is reflected in the list counter', () => {
    const wrapper = factory({
      propsData: { header, list },
      localVue
    })
    const baseListItems = wrapper.findAll(BaseListItem)
    expect(baseListItems.length).toEqual(1 + list.length)
    const avatarWrapper = wrapper.find('div.headline :first-child')
    expect(avatarWrapper.text()).toMatch(list.length + '')
  })

  // not updating synchronously because of [this](https://github.com/vuejs/vue-test-utils/issues/676)
  // nextTick also not working. Had to use setTimeout.
  it('updates list items and count when list prop is updated', () => {
    let listCopy = list.slice(1)
    const wrapper = factory({
      propsData: { header, list: listCopy },
      localVue
      // sync: false
    })

    let baseListItems = wrapper.findAll(BaseListItem)
    expect(baseListItems.length).toEqual(1 + listCopy.length)
    let avatarWrapper = wrapper.find('div.headline :first-child')
    expect(avatarWrapper.text()).toMatch(listCopy.length + '')
    // add new element
    listCopy.push({
      '_id': '29',
      'title': 'founo',
      'desc': 'pa lafapnet te safaw mitlevamu bujufput so cora kokwegti pipwos',
      'tags': ['ceto', 'pumagop'],
      'type': 'format_list_bulleted'
    })

    // return Vue.nextTick().then(() => {
    setTimeout(() => {
      baseListItems = wrapper.findAll(BaseListItem)
      expect(baseListItems.length).toEqual(1 + listCopy.length)
      avatarWrapper = wrapper.find('div.headline :first-child')
      expect(avatarWrapper.text()).toMatch(listCopy.length + '')
    }, 0)
    // })
  })

  it('emits "bl:click" with that item\'s payload on clicking the item', () => {
    const wrapper = factory({
      propsData: { header, list },
      localVue
    }, true)

    const baseListItems = wrapper.findAll(BaseListItem)
    const idx = Math.random() * (baseListItems.length - 1) >> 0
    const someListItem = baseListItems.at(idx)

    someListItem.trigger('click')

    const emitObj = wrapper.emitted('bl:click')
    expect(emitObj).toBeTruthy()
    expect(emitObj[0][0]).toEqual(list[idx])
  })

  // same issue child componenent not mounting
  it('emits "bl:add" on clicking the last list item', () => {
    const wrapper = factory({
      propsData: { header, list },
      localVue
    }, true)

    const listOfItemWrappers = wrapper.findAll(BaseListItem)
    const lastItem = listOfItemWrappers.at(listOfItemWrappers.length - 1)

    lastItem.trigger('click')

    const emitObj = wrapper.emitted('bl:add')
    expect(emitObj).toBeTruthy()
  })

  it('should match snapshot', () => {
    const wrapper = factory({
      propsData: { header, list },
      stubs: {
        'base-list-item': '<div></div>'
      }
    }, true)

    expect(wrapper.html()).toMatchSnapshot()
  })
})
