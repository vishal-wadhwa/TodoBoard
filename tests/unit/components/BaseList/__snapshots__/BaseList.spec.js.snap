// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseList.vue should match snapshot 1`] = `
<div class="container fluid">
  <div class="layout mx-1"><span class="v-badge text--fit"> <span class="headline font-weight-light">Listxyz</span><span class="v-badge__badge black"><span>3</span></span></span>
    <div class="spacer"></div> <button type="button" class="v-btn v-btn--flat v-btn--icon v-btn--small theme--light">
      <div class="v-btn__content"><i aria-hidden="true" class="v-icon material-icons theme--light" style="font-size: 16px;">close</i></div>
    </button>
  </div>
  <div title="sanmu" type="list" desc="oh cunihnon vocew lab sirrufoka loj ujo utceb lez ej" tags="fa,jebpu" ribbon-color="black" view="full"></div>
  <div title="iz" type="playlist_play" desc="vuha ogpufub in jufwozi mira ep po gozseic burni zictam" tags="sikgaok,co" ribbon-color="black" view="full"></div>
  <div title="duzjukve" desc="giv howseb ikopir lo vehladdih delolir uj dafassu joig mocun" tags="magew,wobsudig" ribbon-color="black" view="full"></div>
  <div view="new" title="Create new"></div>
</div>
`;
